sound/soc/sof/virtio-fe.c

```diff
@@ -0,0 +1,734 @@
+// SPDX-License-Identifier: (GPL-2.0 OR BSD-3-Clause)
+/*
+ * This file is provided under a dual BSD/GPLv2 license.  When using or
+ * redistributing this file, you may do so under either license.
+ *
+ * Copyright(c) 2017-2019 Intel Corporation. All rights reserved.
+ *
+ * Author:	Libin Yang <libin.yang@intel.com>
+ *		Luo Xionghu <xionghu.luo@intel.com>
+ *		Liam Girdwood <liam.r.girdwood@linux.intel.com>
+ *		Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>
+ */
+
+/*
+ * virt IO FE driver
+ *
+ * The SOF driver thinks this driver is another audio DSP, however the calls
+ * made by the SOF driver core do not directly go to HW, but over a virtIO
+ * message Q to the virtIO BE driver.
+ *
+ * The virtIO message Q uses the *exact* same IPC structures as we currently
+ * use in the mailbox.
+ *
+ * The mailbox IO and TX/RX msg functions below do IO on the virt IO Q.
+ */
+
+#include <linux/completion.h>
+#include <linux/device.h>
+#include <linux/firmware.h>
+#include <linux/interrupt.h>
+#include <linux/module.h>
+#include <linux/scatterlist.h>
+#include <linux/virtio.h>
+#include <linux/virtio_config.h>
+#include <linux/virtio_ids.h>
+#include <linux/virtio_ring.h>
+#include <sound/sof.h>
+#include <sound/sof/virtio.h>
+
+#include "ops.h"
+#include "sof-priv.h"
+
+#define SOF_COMPONENT_NAME	"sof-audio-component"
+#define DATA_TIMEOUT_MS		600
+
+static const char *const sof_vq_names[SOF_VIRTIO_NUM_OF_VQS] = {
+	SOF_VIRTIO_IPC_CMD_VQ_NAME,
+	SOF_VIRTIO_IPC_PSN_VQ_NAME,
+	SOF_VIRTIO_DATA_VQ_NAME,
+};
+
+struct sof_vfe {
+	struct snd_sof_dev *sdev;
+
+	/* IPC cmd from frontend to backend */
+	struct virtqueue *ipc_cmd_vq;
+
+	/* IPC position update from backend to frontend */
+	struct virtqueue *ipc_psn_vq;
+
+	/* audio data in both directions */
+	struct virtqueue *data_vq;
+
+	/* position update work */
+	struct work_struct posn_update_work;
+
+	/* current pending cmd message */
+	struct snd_sof_ipc_msg *msg;
+
+	/* current and pending notification */
+	struct snd_sof_ipc_msg *not;
+	struct sof_ipc_stream_posn posn;
+
+	struct sof_vfe_ipc_tplg_resp tplg;
+
+	struct completion completion;
+
+	/* A shared capture / playback virtual queue data buffer */
+	union {
+		struct dsp_sof_data_req data_req;
+		struct dsp_sof_data_resp data_resp;
+	};
+
+	/* Headers, used as a playback response or capture request */
+	union {
+		u8 hdr_req[HDR_SIZE_REQ];
+		u8 hdr_resp[HDR_SIZE_RESP];
+	};
+};
+
+/*
+ * IPC Firmware ready.
+ */
+static int sof_vfe_fw_ready(struct snd_sof_dev *sdev, u32 msg_id)
+{
+	return 0;
+};
+
+/* used to send IPC to BE */
+static int sof_vfe_send_msg(struct snd_sof_dev *sdev,
+			    struct snd_sof_ipc_msg *msg)
+{
+	struct sof_vfe *vfe = sdev->pdata->vfe;
+	int ret;
+	struct scatterlist sg_out, sg_in, *sgs[] = {&sg_out, &sg_in};
```

Comment:
one line per variable please...
this is really hard to visualize.

Reply:
@plbossart let's compromise: I'll split it into two:
```
	struct scatterlist sg_out, sg_in;
	struct scatterlist *sgs[] = {&sg_out, &sg_in};
```